### PR TITLE
Parse Expressions to Filter

### PR DESCRIPTION
**Pull request by b4l**
It's far from the complete Mapfile spec but should cover most of the swisstopo use cases.

### CODE REVIEW

File: src/MapfileStyleParser.ts [hunk 1]
```diff
@@ -48,22 +55,232 @@ export class MapfileStyleParser implements StyleParser {
...
+      return ['FN_strMatches', classItem, expression];
+      break;
+    case '{':
+      return [
+        'FN_strMatches',
+        classItem,
+        `/${expression.substring(1, expression.length - 1).replace(',', '|')}/`,
+      ];
+    default:
+      break;
```
Comment (ger-benjamin) [inline src/MapfileStyleParser.ts:80]: You can remove the default as it does nothing.

File: src/MapfileStyleParser.ts [hunk 2]
```diff
@@ -48,22 +55,232 @@ export class MapfileStyleParser implements StyleParser {
...
+  getFilterFromClassItem(mapfileClass: any): Filter | undefined {
+    const classItem: string = mapfileClass.classitem;
+    assert(mapfileClass.classitem, 'Mapfile CLASSITEM undefined!');
+    const expression: string = mapfileClass.expression;
+    switch (expression.charAt(0)) {
+    case '"':
+      return ['==' as ComparisonOperator, mapfileClass.classitemclassItem, expression];
+    case '/':
+      return ['FN_strMatches', classItem, expression];
+      break;
```
Comment (ger-benjamin) [inline src/MapfileStyleParser.ts:72]: It will never reach the break. You can remove it

File: src/MapfileStyleParser.ts (hunk 1 above)
Comment (ger-benjamin) [reply src/MapfileStyleParser.ts:80]: Check also other switch on this file please

File: src/MapfileStyleParser.ts [hunk 3]
```diff
@@ -48,22 +55,232 @@ export class MapfileStyleParser implements StyleParser {
   }
 
   /**
-   * Get the GeoStyler-Style Filter from an Mapfile Expression.
+   * Get the GeoStyler-Style Filter from an Mapfile EXPRESSION and CLASSITEM.
+   *
+   * @param {object} mapfileClass The Mapfile Class
+   * @return {Filter} The GeoStyler-Style Filter
+   */
+  getFilterFromClassItem(mapfileClass: any): Filter | undefined {
```
[the remark this answers is not in the document]
Comment (b4l) [reply src/MapfileStyleParser.ts:63]: It's actually an error as it should always return a Filter.

[the remark this answers is not in the document]
Comment (ger-benjamin) [reply src/MapfileStyleParser.ts:63]: In this case, the return type must be `Filter`. undefined is not possible.
And throw an error if it's not possible to throw a Filter.

File: src/MapfileStyleParser.ts (hunk 1 above)
Comment (b4l) [reply src/MapfileStyleParser.ts:80]: Ok, I removed the empty default cases and set the linting to off.

File: src/MapfileStyleParser.ts [hunk 4]
```diff
@@ -48,22 +55,232 @@ export class MapfileStyleParser implements StyleParser {
...
-    return;
+    // capture nested expressions
+    if (/^\(/.test(mapfileExpression)) {
+      // split up expression
+      const nestedExpressions: Array<string> = [];
+      let index = 0;
+      let fromIndex = 0;
+      let parantheseCount = 0;
+
+      for (const char of mapfileExpression) {
```
[the remark this answers is not in the document]
Comment (b4l) [reply src/MapfileStyleParser.ts:152]: Property 'forEach' does not exist on type 'string'.ts(2339)

[the remark this answers is not in the document]
Comment (ger-benjamin) [reply src/MapfileStyleParser.ts:152]: Oh that's true, I was thinking that was an array of char.
In this case, leave it like that or do mapfileExpression.split('').forEach(char, index) => {...}.
As you want.

File: src/MapfileStyleParser.ts [hunk 5]
```diff
@@ -48,22 +55,232 @@ export class MapfileStyleParser implements StyleParser {
...
+   * Get the GeoStyler-Style Filter from an Mapfile EXPRESSION.
    *
    * @param {string} mapfileExpression The Mapfile Expression
    * @return {Filter} The GeoStyler-Style Filter
    */
-  getFilterFromExpression(mapfileExpression: string): Filter | undefined {
-    // TODO: expression tokenizer, lexer, parser ???
-    if (mapfileExpression) {
-      console.warn(`Not able to parse expression: ${mapfileExpression}`);
+  getFilterFromMapfileExpression(mapfileExpression: string): Filter {
```
[the remark this answers is not in the document]
Comment (b4l) [reply src/MapfileStyleParser.ts:136]: Split out the expression split.

File: src/MapfileStyleParser.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (b4l) [reply src/MapfileStyleParser.ts]: I kind of like it without split

**Comment by b4l** [comment]
Thank you for the review!

I know TDD. I am open to integrate this practice. Maybe after merging this #4 and #5 ?

Typing would be nice indeed though not a pressing issue for me. I was thinking about addressing the Symbols and Labels first to reach MVP as soon as possible.

**Pull request merged**